File: parallelthreadring/src/parallelthreadring.erl
%% Compile under Linux and macOS with "erlc -W  -o parallelthreadring.beam  parallelthreadring.erl"
%% Run with "time erl -noshell -run parallelthreadring start -s init stop -hops 10000 -nodes 1000 -tokens 10"
%%
%% Asynchronous message passing is used for passing the token between the node Erlang process.
%% Each node continuously accepts tokens; if the token is larger than 1,
%% it passes the token minus 1 on, otherwise it sends "done" message to the main process (when token == 0).
%%
%% In Erlang, a program terminates when the main process terminates.

-module(parallelthreadring).
-export([start/0]).

%% @doc Send M messages through a ring of N processes.
%% nth

spawn_nodes(0, _, _) ->
    [];
spawn_nodes(N, Main, Hops) ->
    [spawn_link(fun () -> nodes(N, Main, Hops) end) | spawn_nodes(N - 1, Main, Hops)].

init_nodes(0, [], [])->
    skip;
init_nodes(Nodes, [P|Ps], [N|Ns])->
    P ! {nextp, N},
    init_nodes(Nodes - 1, Ps, Ns).

start_nodes(0, _, _)->
    skip;
start_nodes(N, [P|_]=Ps, Hops)->
    P ! Hops,
    start_nodes(N - 1, Ps, Hops).    

join_nodes(0)->
    skip;
join_nodes(N)->
    receive
        done ->
            join_nodes(N-1)
    end.
    
nextNode([H|Hs])->
    lists:append(Hs, [H]).
    
start() ->
  {ok, [H | _]} = init:get_argument(hops),
  {ok, [N | _]} = init:get_argument(nodes),
  {ok, [T | _]} = init:get_argument(tokens),
  {Hops, _} = string:to_integer(H),
  {Nodes, _} = string:to_integer(N),
  {Tokens, _} = string:to_integer(T),
  Main = self(),
  Ps = spawn_nodes(Nodes, Main, Hops),
  PPs = nextNode(Ps),
  init_nodes(Nodes, Ps, PPs),
  start_nodes(Tokens, Ps, Hops),
  join_nodes(Tokens).

nodes(Next, Main)->
    receive
        1 ->
            %%io:format("done~p~n", [Next]),
            Main ! done,
            nodes(Next, Main);
        Index ->
            Next ! Index - 1,
            nodes(Next, Main)
    end.
    
    
nodes(Id, Main, Hops) ->
  receive
    {nextp, Next} ->
        nodes(Next, Main)
  end.

File: parallelthreadring/src/parallelthreadring.java
/* Compile under Linux and macOS with "javac threadring.java"; run with
"java ThreadRing"

Java monitors with signalling by notify and wait are used. Class Node defines
active objects in the sense that they have a method that can be called and each
has a thread that calls another object. The thread and the method of each object
are synchronized, i.e. run mutually exclusive. Each object passes the token by
calling the pass method of the next object. That method notifies the object's
thread to pass on the token to its next object. A token value of -1 is used by
the thread to check if a valid token has already been passed or the thread needs
to wait for it.

In Java, a program terminates when all threads terminate. Each node thread
terminates when it won't receive any more tokens, which is when the received
token is less than the number of nodes.
*/

import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class Node extends Thread {
  final Lock lock = new ReentrantLock();
  final Condition empty = lock.newCondition();
  final Condition full = lock.newCondition();
  Node next;
  int token = - 1;
  int n;
  Counter c;
  Node(int n, Counter c) {
    this.n = n; this.c = c;
    setDaemon(true); start();
  }
  void pass(int t) {
    lock.lock();
    try {
      while (token >= 0) empty.await();
      token = t; full.signal();
    } catch (InterruptedException e) {
    } finally {lock.unlock();
    }
  }
  public void run() {
    lock.lock();
    try {
      while (true) {
        while (token == - 1) full.await();
        // System.out.println(n + " received " + token);
        if (token > 0) next.pass(token - 1);
        else c.dec();
        token = - 1; empty.signal();
      }
    } catch (InterruptedException e) {
    //} finally {lock.unlock();
    }
  }
}
class Counter {
  int count;
  Counter(int c) {count = c;}
  synchronized void dec() {count -= 1; if (count == 0) notify();}
  synchronized void zero() {
    if (count > 0)
      try {wait();} catch (InterruptedException e) {}
  }
}
class ParallelThreadRing {
  static final int Hops = 10000;
  static final int Nodes = 1000;
  static final int Tokens = 999;
  public static void main(String args[]) {
    Counter c = new Counter(Tokens);
    Node nd0 = new Node(0, c);
    Node nd = nd0;
    for (int n = 1; n < Nodes; n++) {
      nd.next = new Node(n, c); nd = nd.next;
    }
    nd.next = nd0;
    for (int t = 0; t < Tokens; t++) nd0.pass(Hops);
    c.zero();
  }
}
